servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,619 @@
+// TODO: The fieldwise check is unimplemented, but still this one is useful, as a strong annotation about the needs.
+#define ARRAYS_COMPATIBLE_FIELDWISE(m_type_a, m_type_b) ARRAYS_COMPATIBLE(m_type_a, m_type_b)
+
+// Design principles:
+// - Very little validation is done, and normally only in dev or debug builds.
+// - Error reporting is generally simple: returning an id of 0 or a false boolean.
+// - Certain enums/constants/structs follow Vulkan values/layout. That makes things easier for RDDVulkan (it asserts compatibility).
+// - We allocate as little as possible in functions expected to be quick (a counterexample is loading/saving shaders) and use alloca() whenever suitable.
+// - We try to back our various ids with the native ones. When a bespoke id is needed, if the address of the object is fixed (e.g., a RBSet item), a pointer to it is under the hood.
+// - Every struct has default initializers.
+// - Using VectorView to take array-like arguments. Vector<uint8_t> is an exception (an indiom for "BLOB").
+class RenderingDeviceDriver : public RenderingDeviceCommons {
+public:
+	// Id types declared befor anything else to prevent cyclic dependencies between the different concerns.
+
+	// BUFFERS.
+	using BufferID = uint64_t;
+	// TEXTURE.
+	using TextureID = uint64_t;
+	// SAMPLER.
+	using SamplerID = uint64_t;
+	// VERTEX ARRAY.
+	using VertexFormatID = uint64_t;
+	// COMMAND BUFFERS.
+	using CommandPoolID = uint64_t;
+	using CommandBufferID = uint64_t;
+	// FRAMEBUFFER.
+	using FramebufferID = uint64_t;
+	// SHADER.
+	using ShaderID = uint64_t;
+	// UNIFORM SET.
+	using UniformSetID = uint64_t;
+	// RENDERING.
+	using RenderPassID = uint64_t;
+	using PipelineID = uint64_t;
+	// COMPUTE.
+	using PipelineID = uint64_t;
+	// QUERIES.
+	using QueryPoolID = uint64_t;
+
+	/****************/
+	/**** MEMORY ****/
+	/****************/
+
+	enum MemoryAllocationType {
+		MEMORY_ALLOCATION_TYPE_CPU, // For images, CPU allocation also means linear, GPU is tiling optimal.
+		MEMORY_ALLOCATION_TYPE_GPU,
+	};
+
+	/*****************/
+	/**** BUFFERS ****/
+	/*****************/
+
+	enum BufferUsageBits {
+		BUFFER_USAGE_TRANSFER_FROM_BIT = (1 << 0),
+		BUFFER_USAGE_TRANSFER_TO_BIT = (1 << 1),
+		BUFFER_USAGE_TEXEL_BIT = (1 << 2),
+		BUFFER_USAGE_UNIFORM_BIT = (1 << 4),
+		BUFFER_USAGE_STORAGE_BIT = (1 << 5),
+		BUFFER_USAGE_INDEX_BIT = (1 << 6),
+		BUFFER_USAGE_VERTEX_BIT = (1 << 7),
+		BUFFER_USAGE_INDIRECT_BIT = (1 << 8),
+	};
+
+	virtual BufferID buffer_create(uint64_t p_size, BitField<BufferUsageBits> p_usage, MemoryAllocationType p_allocation_type) = 0;
+	// Only for a buffer with BUFFER_USAGE_TEXEL_BIT.
+	virtual bool buffer_set_texel_format(BufferID p_buffer, DataFormat p_format) = 0;
+	virtual void buffer_free(BufferID p_buffer) = 0;
+	virtual uint64_t buffer_get_allocation_size(BufferID p_buffer) = 0;
+	virtual uint8_t *buffer_map(BufferID p_buffer) = 0;
+	virtual void buffer_unmap(BufferID p_buffer) = 0;
```

Comment:
Probably better to keep it simple, I don't think its needed.

Reply:
As with other features in the driver, I've generally chosen to expose what is currently actually used instead of opening dorrs to the whole functionality. This is something that future maintainers (maybe yourself) will have to change in case there's an actual use case for partial mapping.

Reply:
For reference, Metal doesn't have to map / unmap buffers. `MTLBuffer` just has a pointer to the buffer, and you must respect standard memory coherency rules.